helper/resource/teststep_providers_test.go

```diff
@@ -342,6 +344,66 @@ func TestTest_TestStep_ExternalProviders_To_ProviderFactories_StateUpgraders(t *
 	})
 }
 
+func TestTest_TestStep_Taint(t *testing.T) {
+	t.Parallel()
+
+	var idOne, idTwo string
+
+	Test(t, TestCase{
+		ExternalProviders: map[string]ExternalProvider{
+			"random": {
+				Source:            "registry.terraform.io/hashicorp/random",
+				VersionConstraint: "3.3.2",
+			},
+		},
+		Steps: []TestStep{
+			{
+				Config: `resource "random_string" "test" {
+					length = 10
+				}`,
+				Check: ComposeAggregateTestCheckFunc(
+					TestCheckResourceAttr("random_string.test", "length", "10"),
+					ExtractResourceAttr("random_string.test", "id", &idOne),
+				),
+			},
+			{
+				Taint: []string{"random_string.test"},
+				Config: `resource "random_string" "test" {
+					length = 10
+				}`,
+				Check: ComposeAggregateTestCheckFunc(
+					TestCheckResourceAttr("random_string.test", "length", "10"),
+					ExtractResourceAttr("random_string.test", "id", &idTwo),
+				),
+			},
+		},
```

Comment:
Rather than requiring the testing to lookup, download, and install an external provider, we can inline create one, e.g.

```suggestion
		ProviderFactories: map[string]func() (*schema.Provider, error){
			"random": func() (*schema.Provider, error) { //nolint:unparam // required signature
				return &schema.Provider{
					ResourcesMap: map[string]*schema.Resource{
						"random_id": {
							CreateContext: func(_ context.Context, d *schema.ResourceData, _ interface{}) diag.Diagnostics {
								d.SetId(time.Now().String())
								return nil
							},
							DeleteContext: func(_ context.Context, _ *schema.ResourceData, _ interface{}) diag.Diagnostics {
								return nil
							},
							ReadContext: func(_ context.Context, _ *schema.ResourceData, _ interface{}) diag.Diagnostics {
								return nil
							},
							Schema: map[string]*schema.Schema{},
						},
					},
				}, nil
			},
		},
		Steps: []TestStep{
			{
				Config: `resource "random_id" "test" {}`,
				Check: ComposeAggregateTestCheckFunc(
					extractResourceAttr("random_string.test", "id", &idOne),
				),
			},
			{
				Taint: []string{"random_id.test"},
				Config: `resource "random_id" "test" {}`,
				Check: ComposeAggregateTestCheckFunc(
					extractResourceAttr("random_string.test", "id", &idTwo),
				),
			},
		},
```